// @react-spectrum/color/dist/main.js
require("./main.css");
var $aAzec$reactspectrumutils = require("@react-spectrum/utils");
var $aAzec$reactariautils = require("@react-aria/utils");
var $aAzec$react = require("react");
var $aAzec$reactariacolor = require("@react-aria/color");
var $aAzec$reactstatelycolor = require("@react-stately/color");
var $aAzec$reactariafocus = require("@react-aria/focus");
var $aAzec$reactspectrumprovider = require("@react-spectrum/provider");
var $aAzec$reactspectrumlayout = require("@react-spectrum/layout");
var $aAzec$reactspectrumlabel = require("@react-spectrum/label");
var $aAzec$reactariainteractions = require("@react-aria/interactions");
var $aAzec$reactariai18n = require("@react-aria/i18n");
var $aAzec$reactspectrumtextfield = require("@react-spectrum/textfield");

function $parcel$exportWildcard(dest, source) {
  Object.keys(source).forEach(function(key) {
    if (key === 'default' || key === '__esModule' || dest.hasOwnProperty(key)) {
      return;
    }

    Object.defineProperty(dest, key, {
      enumerable: true,
      get: function get() {
        return source[key];
      }
    });
  });

  return dest;
}
function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $5f016e950c2dc3a6$exports = {};

$parcel$export($5f016e950c2dc3a6$exports, "ColorArea", () => $5f016e950c2dc3a6$export$b2103f68a961418e);



var $f61d632765728b2e$exports = {};

$parcel$export($f61d632765728b2e$exports, "spectrum-ColorHandle", () => $f61d632765728b2e$export$cd9afaa621b6216f, (v) => $f61d632765728b2e$export$cd9afaa621b6216f = v);
$parcel$export($f61d632765728b2e$exports, "is-focused", () => $f61d632765728b2e$export$e7dc768d35940237, (v) => $f61d632765728b2e$export$e7dc768d35940237 = v);
$parcel$export($f61d632765728b2e$exports, "focus-ring", () => $f61d632765728b2e$export$f39a09f249340e2a, (v) => $f61d632765728b2e$export$f39a09f249340e2a = v);
$parcel$export($f61d632765728b2e$exports, "is-disabled", () => $f61d632765728b2e$export$d35bc1e505d1ebbf, (v) => $f61d632765728b2e$export$d35bc1e505d1ebbf = v);
$parcel$export($f61d632765728b2e$exports, "spectrum-ColorHandle-color", () => $f61d632765728b2e$export$afe4c366ed4e659c, (v) => $f61d632765728b2e$export$afe4c366ed4e659c = v);
var $f61d632765728b2e$export$cd9afaa621b6216f;
var $f61d632765728b2e$export$e7dc768d35940237;
var $f61d632765728b2e$export$f39a09f249340e2a;
var $f61d632765728b2e$export$d35bc1e505d1ebbf;
var $f61d632765728b2e$export$afe4c366ed4e659c;
$f61d632765728b2e$export$cd9afaa621b6216f = "spectrum-ColorHandle_5a9f41";
$f61d632765728b2e$export$e7dc768d35940237 = "is-focused_5a9f41";
$f61d632765728b2e$export$f39a09f249340e2a = "focus-ring_5a9f41";
$f61d632765728b2e$export$d35bc1e505d1ebbf = "is-disabled_5a9f41";
$f61d632765728b2e$export$afe4c366ed4e659c = "spectrum-ColorHandle-color_5a9f41";


var $f65977ae56facfaa$exports = {};

$parcel$export($f65977ae56facfaa$exports, "spectrum-ColorLoupe", () => $f65977ae56facfaa$export$88d6ed2576b21809, (v) => $f65977ae56facfaa$export$88d6ed2576b21809 = v);
$parcel$export($f65977ae56facfaa$exports, "is-open", () => $f65977ae56facfaa$export$a9781837241c946d, (v) => $f65977ae56facfaa$export$a9781837241c946d = v);
$parcel$export($f65977ae56facfaa$exports, "spectrum-ColorLoupe-outer", () => $f65977ae56facfaa$export$d6ad38291e8aca9c, (v) => $f65977ae56facfaa$export$d6ad38291e8aca9c = v);
$parcel$export($f65977ae56facfaa$exports, "spectrum-ColorLoupe-inner-background", () => $f65977ae56facfaa$export$397af82b48750fe5, (v) => $f65977ae56facfaa$export$397af82b48750fe5 = v);
$parcel$export($f65977ae56facfaa$exports, "spectrum-ColorLoupe-inner-checker", () => $f65977ae56facfaa$export$f9a2a02e0b271024, (v) => $f65977ae56facfaa$export$f9a2a02e0b271024 = v);
var $f65977ae56facfaa$export$88d6ed2576b21809;
var $f65977ae56facfaa$export$a9781837241c946d;
var $f65977ae56facfaa$export$d6ad38291e8aca9c;
var $f65977ae56facfaa$export$397af82b48750fe5;
var $f65977ae56facfaa$export$f9a2a02e0b271024;
$f65977ae56facfaa$export$88d6ed2576b21809 = "spectrum-ColorLoupe_c818a8";
$f65977ae56facfaa$export$a9781837241c946d = "is-open_c818a8";
$f65977ae56facfaa$export$d6ad38291e8aca9c = "spectrum-ColorLoupe-outer_c818a8";
$f65977ae56facfaa$export$397af82b48750fe5 = "spectrum-ColorLoupe-inner-background_c818a8";
$f65977ae56facfaa$export$f9a2a02e0b271024 = "spectrum-ColorLoupe-inner-checker_c818a8";



function $2b9adc23f6c7cae1$export$a3cc47cee1c1ccc(props) {
    let { value: value , isDisabled: isDisabled , isDragging: isDragging , isFocused: isFocused , children: children , className: className = '' , ...otherProps } = props;
    let patternId = $aAzec$reactariautils.useId();
    let valueCSS = value.toString('css');
    return(/*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f61d632765728b2e$exports)), 'spectrum-ColorHandle', {
            'is-focused': isFocused,
            'is-disabled': isDisabled
        }) + ' ' + className,
        ...otherProps
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f61d632765728b2e$exports)), 'spectrum-ColorHandle-color'),
        style: {
            backgroundColor: valueCSS
        }
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("svg", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe', {
            'is-open': isDragging
        }),
        "aria-hidden": "true"
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("pattern", {
        id: patternId,
        x: "0",
        y: "0",
        width: "16",
        height: "16",
        patternUnits: "userSpaceOnUse"
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("rect", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe-inner-background'),
        x: "0",
        y: "0",
        width: "16",
        height: "16"
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("rect", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe-inner-checker'),
        x: "0",
        y: "0",
        width: "8",
        height: "8"
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("rect", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe-inner-checker'),
        x: "8",
        y: "8",
        width: "8",
        height: "8"
    })), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("path", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe-inner'),
        d: "M25 1a24 24 0 0124 24c0 16.255-24 40-24 40S1 41.255 1 25A24 24 0 0125 1z",
        fill: `url(#${patternId})`
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("path", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe-inner'),
        d: "M25 1a24 24 0 0124 24c0 16.255-24 40-24 40S1 41.255 1 25A24 24 0 0125 1z",
        fill: valueCSS
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("path", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($f65977ae56facfaa$exports)), 'spectrum-ColorLoupe-outer'),
        d: "M25 3A21.98 21.98 0 003 25c0 6.2 4 14.794 11.568 24.853A144.233 144.233 0 0025 62.132a144.085 144.085 0 0010.4-12.239C42.99 39.816 47 31.209 47 25A21.98 21.98 0 0025 3m0-2a24 24 0 0124 24c0 16.255-24 40-24 40S1 41.255 1 25A24 24 0 0125 1z"
    })), children));
}




var $d4b4e0a63f9ca8c1$exports = {};

$parcel$export($d4b4e0a63f9ca8c1$exports, "spectrum-ColorArea", () => $d4b4e0a63f9ca8c1$export$bb6257a55a3c1efc, (v) => $d4b4e0a63f9ca8c1$export$bb6257a55a3c1efc = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "is-focused", () => $d4b4e0a63f9ca8c1$export$e7dc768d35940237, (v) => $d4b4e0a63f9ca8c1$export$e7dc768d35940237 = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "focus-ring", () => $d4b4e0a63f9ca8c1$export$f39a09f249340e2a, (v) => $d4b4e0a63f9ca8c1$export$f39a09f249340e2a = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "spectrum-ColorArea-handle", () => $d4b4e0a63f9ca8c1$export$d8addbd273c4e00, (v) => $d4b4e0a63f9ca8c1$export$d8addbd273c4e00 = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "is-disabled", () => $d4b4e0a63f9ca8c1$export$d35bc1e505d1ebbf, (v) => $d4b4e0a63f9ca8c1$export$d35bc1e505d1ebbf = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "spectrum-ColorArea-gradient", () => $d4b4e0a63f9ca8c1$export$40686f4fcb8a9916, (v) => $d4b4e0a63f9ca8c1$export$40686f4fcb8a9916 = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "spectrum-ColorArea-slider", () => $d4b4e0a63f9ca8c1$export$7d727cacaa7cea1e, (v) => $d4b4e0a63f9ca8c1$export$7d727cacaa7cea1e = v);
$parcel$export($d4b4e0a63f9ca8c1$exports, "spectrum-ColorHandle-color", () => $d4b4e0a63f9ca8c1$export$afe4c366ed4e659c, (v) => $d4b4e0a63f9ca8c1$export$afe4c366ed4e659c = v);
var $d4b4e0a63f9ca8c1$export$bb6257a55a3c1efc;
var $d4b4e0a63f9ca8c1$export$e7dc768d35940237;
var $d4b4e0a63f9ca8c1$export$f39a09f249340e2a;
var $d4b4e0a63f9ca8c1$export$d8addbd273c4e00;
var $d4b4e0a63f9ca8c1$export$d35bc1e505d1ebbf;
var $d4b4e0a63f9ca8c1$export$40686f4fcb8a9916;
var $d4b4e0a63f9ca8c1$export$7d727cacaa7cea1e;
var $d4b4e0a63f9ca8c1$export$afe4c366ed4e659c;
$d4b4e0a63f9ca8c1$export$bb6257a55a3c1efc = "spectrum-ColorArea_35e2c0";
$d4b4e0a63f9ca8c1$export$e7dc768d35940237 = "is-focused_35e2c0";
$d4b4e0a63f9ca8c1$export$f39a09f249340e2a = "focus-ring_35e2c0";
$d4b4e0a63f9ca8c1$export$d8addbd273c4e00 = "spectrum-ColorArea-handle_35e2c0";
$d4b4e0a63f9ca8c1$export$d35bc1e505d1ebbf = "is-disabled_35e2c0";
$d4b4e0a63f9ca8c1$export$40686f4fcb8a9916 = "spectrum-ColorArea-gradient_35e2c0";
$d4b4e0a63f9ca8c1$export$7d727cacaa7cea1e = "spectrum-ColorArea-slider_35e2c0";
$d4b4e0a63f9ca8c1$export$afe4c366ed4e659c = "spectrum-ColorHandle-color_35e2c0";






function $5f016e950c2dc3a6$var$ColorArea(props, ref) {
    props = $aAzec$reactspectrumprovider.useProviderProps(props);
    let { isDisabled: isDisabled  } = props;
    let size = props.size && $aAzec$reactspectrumutils.dimensionValue(props.size);
    let { styleProps: styleProps  } = $aAzec$reactspectrumutils.useStyleProps(props);
    let inputXRef = $aAzec$react.useRef();
    let inputYRef = $aAzec$react.useRef();
    let containerRef = $aAzec$reactspectrumutils.useFocusableRef(ref, inputXRef);
    let state = $aAzec$reactstatelycolor.useColorAreaState(props);
    let { colorAreaProps: colorAreaProps , gradientProps: gradientProps , xInputProps: xInputProps , yInputProps: yInputProps , thumbProps: thumbProps  } = $aAzec$reactariacolor.useColorArea({
        ...props,
        inputXRef: inputXRef,
        inputYRef: inputYRef,
        containerRef: containerRef
    }, state);
    let { focusProps: focusProps , isFocusVisible: isFocusVisible  } = $aAzec$reactariafocus.useFocusRing();
    return(/*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        ...colorAreaProps,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($d4b4e0a63f9ca8c1$exports)), 'spectrum-ColorArea', {
            'is-disabled': isDisabled
        }, styleProps.className),
        ref: containerRef,
        style: {
            ...colorAreaProps.style,
            ...styleProps.style,
            // Workaround around https://github.com/adobe/spectrum-css/issues/1032
            width: size,
            height: size
        }
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        ...gradientProps,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($d4b4e0a63f9ca8c1$exports)), 'spectrum-ColorArea-gradient')
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($2b9adc23f6c7cae1$export$a3cc47cee1c1ccc, {
        value: state.getDisplayColor(),
        isFocused: isFocusVisible,
        isDisabled: isDisabled,
        isDragging: state.isDragging,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($d4b4e0a63f9ca8c1$exports)), 'spectrum-ColorArea-handle'),
        ...thumbProps
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        role: "presentation"
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("input", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($d4b4e0a63f9ca8c1$exports)), 'spectrum-ColorArea-slider'),
        ...$aAzec$reactariautils.mergeProps(xInputProps, focusProps),
        ref: inputXRef
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("input", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($d4b4e0a63f9ca8c1$exports)), 'spectrum-ColorArea-slider'),
        ...$aAzec$reactariautils.mergeProps(yInputProps, focusProps),
        ref: inputYRef
    })))));
}
let $5f016e950c2dc3a6$export$b2103f68a961418e = /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).forwardRef($5f016e950c2dc3a6$var$ColorArea);


var $63a2864687444ae1$exports = {};

$parcel$export($63a2864687444ae1$exports, "ColorWheel", () => $63a2864687444ae1$export$f80663f808113381);



var $8fc6b1a2f4e7f682$exports = {};

$parcel$export($8fc6b1a2f4e7f682$exports, "spectrum-ColorWheel", () => $8fc6b1a2f4e7f682$export$b70a5ca28702fbc6, (v) => $8fc6b1a2f4e7f682$export$b70a5ca28702fbc6 = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "is-focused", () => $8fc6b1a2f4e7f682$export$e7dc768d35940237, (v) => $8fc6b1a2f4e7f682$export$e7dc768d35940237 = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "spectrum-ColorWheel-handle", () => $8fc6b1a2f4e7f682$export$4cd23920eb8ac056, (v) => $8fc6b1a2f4e7f682$export$4cd23920eb8ac056 = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "spectrum-ColorWheel-slider", () => $8fc6b1a2f4e7f682$export$f476ddda713d62b, (v) => $8fc6b1a2f4e7f682$export$f476ddda713d62b = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "is-disabled", () => $8fc6b1a2f4e7f682$export$d35bc1e505d1ebbf, (v) => $8fc6b1a2f4e7f682$export$d35bc1e505d1ebbf = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "is-dragged", () => $8fc6b1a2f4e7f682$export$8778c911bed6c759, (v) => $8fc6b1a2f4e7f682$export$8778c911bed6c759 = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "spectrum-ColorWheel-gradient", () => $8fc6b1a2f4e7f682$export$9afaa791d7500a9b, (v) => $8fc6b1a2f4e7f682$export$9afaa791d7500a9b = v);
$parcel$export($8fc6b1a2f4e7f682$exports, "spectrum-ColorWheel-segment", () => $8fc6b1a2f4e7f682$export$ada9fbffdaa4f10, (v) => $8fc6b1a2f4e7f682$export$ada9fbffdaa4f10 = v);
var $8fc6b1a2f4e7f682$export$b70a5ca28702fbc6;
var $8fc6b1a2f4e7f682$export$e7dc768d35940237;
var $8fc6b1a2f4e7f682$export$4cd23920eb8ac056;
var $8fc6b1a2f4e7f682$export$f476ddda713d62b;
var $8fc6b1a2f4e7f682$export$d35bc1e505d1ebbf;
var $8fc6b1a2f4e7f682$export$8778c911bed6c759;
var $8fc6b1a2f4e7f682$export$9afaa791d7500a9b;
var $8fc6b1a2f4e7f682$export$ada9fbffdaa4f10;
$8fc6b1a2f4e7f682$export$b70a5ca28702fbc6 = "spectrum-ColorWheel_31462a";
$8fc6b1a2f4e7f682$export$e7dc768d35940237 = "is-focused_31462a";
$8fc6b1a2f4e7f682$export$4cd23920eb8ac056 = "spectrum-ColorWheel-handle_31462a";
$8fc6b1a2f4e7f682$export$f476ddda713d62b = "spectrum-ColorWheel-slider_31462a";
$8fc6b1a2f4e7f682$export$d35bc1e505d1ebbf = "is-disabled_31462a";
$8fc6b1a2f4e7f682$export$8778c911bed6c759 = "is-dragged_31462a";
$8fc6b1a2f4e7f682$export$9afaa791d7500a9b = "spectrum-ColorWheel-gradient_31462a";
$8fc6b1a2f4e7f682$export$ada9fbffdaa4f10 = "spectrum-ColorWheel-segment_31462a";







const $63a2864687444ae1$var$WHEEL_THICKNESS = 24;
function $63a2864687444ae1$var$ColorWheel(props, ref) {
    props = $aAzec$reactspectrumprovider.useProviderProps(props);
    let { isDisabled: isDisabled  } = props;
    let size = props.size && $aAzec$reactspectrumutils.dimensionValue(props.size);
    let { styleProps: styleProps  } = $aAzec$reactspectrumutils.useStyleProps(props);
    let inputRef = $aAzec$react.useRef(null);
    let containerRef = $aAzec$reactspectrumutils.useFocusableRef(ref, inputRef);
    let [wheelRadius, setWheelRadius] = $aAzec$react.useState(null);
    let [wheelThickness, setWheelThickness] = $aAzec$react.useState($63a2864687444ae1$var$WHEEL_THICKNESS);
    let resizeHandler = $aAzec$react.useCallback(()=>{
        if (containerRef.current) {
            setWheelRadius(containerRef.current.offsetWidth / 2);
            let thickness = window.getComputedStyle(containerRef.current).getPropertyValue('--spectrum-colorwheel-track-thickness');
            if (thickness) setWheelThickness(parseInt(thickness, 10));
        }
    }, [
        containerRef,
        setWheelRadius,
        setWheelThickness
    ]);
    $aAzec$react.useEffect(()=>{
        // the size observer's fallback to the window resize event doesn't fire on mount
        if (wheelRadius == null) resizeHandler();
    }, [
        wheelRadius,
        resizeHandler
    ]);
    $aAzec$reactariautils.useResizeObserver({
        ref: containerRef,
        onResize: resizeHandler
    });
    let state = $aAzec$reactstatelycolor.useColorWheelState(props);
    let { trackProps: trackProps , inputProps: inputProps , thumbProps: thumbProps  } = $aAzec$reactariacolor.useColorWheel({
        ...props,
        innerRadius: wheelRadius - wheelThickness,
        outerRadius: wheelRadius
    }, state, inputRef);
    let { focusProps: focusProps , isFocusVisible: isFocusVisible  } = $aAzec$reactariafocus.useFocusRing();
    return(/*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($8fc6b1a2f4e7f682$exports)), 'spectrum-ColorWheel', {
            'is-disabled': isDisabled
        }, styleProps.className),
        ref: containerRef,
        style: {
            ...styleProps.style,
            // Workaround around https://github.com/adobe/spectrum-css/issues/1032
            // @ts-ignore
            'width': size,
            'height': size
        }
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        ...trackProps,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($8fc6b1a2f4e7f682$exports)), 'spectrum-ColorWheel-gradient')
    }), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($2b9adc23f6c7cae1$export$a3cc47cee1c1ccc, {
        value: state.getDisplayColor(),
        isFocused: isFocusVisible,
        isDisabled: isDisabled,
        isDragging: state.isDragging,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($8fc6b1a2f4e7f682$exports)), 'spectrum-ColorWheel-handle'),
        ...thumbProps
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("input", {
        ...focusProps,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($8fc6b1a2f4e7f682$exports)), 'spectrum-ColorWheel-slider'),
        ...inputProps,
        ref: inputRef
    }))));
}
/**
 * ColorWheels allow users to adjust the hue of an HSL or HSB color value on a circular track.
 */ let $63a2864687444ae1$export$f80663f808113381 = /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).forwardRef($63a2864687444ae1$var$ColorWheel);


var $4537dec0de08c277$exports = {};

$parcel$export($4537dec0de08c277$exports, "ColorSlider", () => $4537dec0de08c277$export$44fd664bcca5b6fb);





var $28060baf29cc5ac7$exports = {};

$parcel$export($28060baf29cc5ac7$exports, "spectrum-ColorSlider", () => $28060baf29cc5ac7$export$a5f8b0989b27a604, (v) => $28060baf29cc5ac7$export$a5f8b0989b27a604 = v);
$parcel$export($28060baf29cc5ac7$exports, "is-focused", () => $28060baf29cc5ac7$export$e7dc768d35940237, (v) => $28060baf29cc5ac7$export$e7dc768d35940237 = v);
$parcel$export($28060baf29cc5ac7$exports, "spectrum-ColorSlider-handle", () => $28060baf29cc5ac7$export$8c273ba92fabe1f1, (v) => $28060baf29cc5ac7$export$8c273ba92fabe1f1 = v);
$parcel$export($28060baf29cc5ac7$exports, "spectrum-ColorSlider-slider", () => $28060baf29cc5ac7$export$2f08b72944af536, (v) => $28060baf29cc5ac7$export$2f08b72944af536 = v);
$parcel$export($28060baf29cc5ac7$exports, "spectrum-ColorSlider-container--horizontal", () => $28060baf29cc5ac7$export$33e97914f6746614, (v) => $28060baf29cc5ac7$export$33e97914f6746614 = v);
$parcel$export($28060baf29cc5ac7$exports, "spectrum-ColorSlider-container--vertical", () => $28060baf29cc5ac7$export$cd1c093289104c66, (v) => $28060baf29cc5ac7$export$cd1c093289104c66 = v);
$parcel$export($28060baf29cc5ac7$exports, "is-disabled", () => $28060baf29cc5ac7$export$d35bc1e505d1ebbf, (v) => $28060baf29cc5ac7$export$d35bc1e505d1ebbf = v);
$parcel$export($28060baf29cc5ac7$exports, "spectrum-ColorSlider--vertical", () => $28060baf29cc5ac7$export$83dcf61b398bfaf1, (v) => $28060baf29cc5ac7$export$83dcf61b398bfaf1 = v);
var $28060baf29cc5ac7$export$a5f8b0989b27a604;
var $28060baf29cc5ac7$export$e7dc768d35940237;
var $28060baf29cc5ac7$export$8c273ba92fabe1f1;
var $28060baf29cc5ac7$export$2f08b72944af536;
var $28060baf29cc5ac7$export$33e97914f6746614;
var $28060baf29cc5ac7$export$cd1c093289104c66;
var $28060baf29cc5ac7$export$d35bc1e505d1ebbf;
var $28060baf29cc5ac7$export$83dcf61b398bfaf1;
$28060baf29cc5ac7$export$a5f8b0989b27a604 = "spectrum-ColorSlider_a5bb06";
$28060baf29cc5ac7$export$e7dc768d35940237 = "is-focused_a5bb06";
$28060baf29cc5ac7$export$8c273ba92fabe1f1 = "spectrum-ColorSlider-handle_a5bb06";
$28060baf29cc5ac7$export$2f08b72944af536 = "spectrum-ColorSlider-slider_a5bb06";
$28060baf29cc5ac7$export$33e97914f6746614 = "spectrum-ColorSlider-container--horizontal_a5bb06";
$28060baf29cc5ac7$export$cd1c093289104c66 = "spectrum-ColorSlider-container--vertical_a5bb06";
$28060baf29cc5ac7$export$d35bc1e505d1ebbf = "is-disabled_a5bb06";
$28060baf29cc5ac7$export$83dcf61b398bfaf1 = "spectrum-ColorSlider--vertical_a5bb06";







function $4537dec0de08c277$var$ColorSlider(props, ref) {
    props = $aAzec$reactspectrumprovider.useProviderProps(props);
    let { isDisabled: isDisabled , channel: channel , orientation: orientation , label: label , showValueLabel: showValueLabel , 'aria-label': ariaLabel  } = props;
    let vertical = orientation === 'vertical';
    let { styleProps: styleProps  } = $aAzec$reactspectrumutils.useStyleProps(props);
    let { locale: locale  } = $aAzec$reactariai18n.useLocale();
    let inputRef = $aAzec$react.useRef();
    let trackRef = $aAzec$react.useRef();
    let domRef = $aAzec$reactspectrumutils.useFocusableRef(ref, inputRef);
    let state = $aAzec$reactstatelycolor.useColorSliderState({
        ...props,
        locale: locale
    });
    // If vertical and a label is provided, use it as an aria-label instead.
    if (vertical && label) {
        ariaLabel = ariaLabel || (typeof label === 'string' ? label : null);
        label = null;
    }
    // If no external label, aria-label or aria-labelledby is provided,
    // default to displaying the localized channel value.
    // Specifically check if label is undefined. If label is `null` then display no visible label.
    // A default aria-label is provided by useColorSlider in that case.
    if (label === undefined && !ariaLabel && !props['aria-labelledby'] && !vertical) label = state.value.getChannelName(channel, locale);
    // Show the value label by default if there is a visible label
    if (showValueLabel == null) showValueLabel = !!label;
    let { inputProps: inputProps , thumbProps: thumbProps , trackProps: trackProps , labelProps: labelProps , outputProps: outputProps  } = $aAzec$reactariacolor.useColorSlider({
        ...props,
        label: label,
        'aria-label': ariaLabel,
        trackRef: trackRef,
        inputRef: inputRef
    }, state);
    let { isFocusVisible: isFocusVisible  } = $aAzec$reactariainteractions.useFocusVisible();
    let [isFocused, setIsFocused] = $aAzec$react.useState(false);
    let { focusProps: focusProps  } = $aAzec$reactariainteractions.useFocus({
        isDisabled: isDisabled,
        onFocusChange: setIsFocused
    });
    let alignLabel;
    if (vertical) alignLabel = 'center';
    else if (label != null && showValueLabel) alignLabel = 'space-between';
    else if (label != null) alignLabel = 'flex-start';
    else if (showValueLabel) alignLabel = 'flex-end';
    return(/*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        ref: domRef,
        ...styleProps,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($28060baf29cc5ac7$exports)), {
            'spectrum-ColorSlider-container--horizontal': !vertical,
            'spectrum-ColorSlider-container--vertical': vertical
        })
    }, label && /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($aAzec$reactspectrumlayout.Flex, {
        direction: "row",
        justifyContent: alignLabel
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($aAzec$reactspectrumlabel.Label, labelProps, label), showValueLabel && /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($aAzec$reactspectrumlabel.Label, {
        elementType: "span"
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("output", outputProps, state.value.formatChannelValue(channel, locale)))), /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("div", {
        ...trackProps,
        ref: trackRef,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($28060baf29cc5ac7$exports)), 'spectrum-ColorSlider', {
            'is-disabled': isDisabled,
            'spectrum-ColorSlider--vertical': vertical
        })
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($2b9adc23f6c7cae1$export$a3cc47cee1c1ccc, {
        value: state.getDisplayColor(),
        isFocused: isFocused && isFocusVisible,
        isDisabled: isDisabled,
        isDragging: state.isThumbDragging(0),
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($28060baf29cc5ac7$exports)), 'spectrum-ColorSlider-handle'),
        ...thumbProps
    }, /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement("input", {
        ...inputProps,
        ...focusProps,
        ref: inputRef,
        className: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($28060baf29cc5ac7$exports)), 'spectrum-ColorSlider-slider')
    })))));
}
/**
 * ColorSliders allow users to adjust an individual channel of a color value.
 */ let $4537dec0de08c277$export$44fd664bcca5b6fb = /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).forwardRef($4537dec0de08c277$var$ColorSlider);


var $8bdb9bc27ff3debb$exports = {};

$parcel$export($8bdb9bc27ff3debb$exports, "ColorField", () => $8bdb9bc27ff3debb$export$b865d4358897bb17);


var $ddec4d6aa21146c4$exports = {};

$parcel$export($ddec4d6aa21146c4$exports, "react-spectrum-ColorField-input", () => $ddec4d6aa21146c4$export$7d80c6630a750b1, (v) => $ddec4d6aa21146c4$export$7d80c6630a750b1 = v);
var $ddec4d6aa21146c4$export$7d80c6630a750b1;
$ddec4d6aa21146c4$export$7d80c6630a750b1 = "react-spectrum-ColorField-input_7bde3a";






function $8bdb9bc27ff3debb$var$ColorField(props, ref) {
    props = $aAzec$reactspectrumprovider.useProviderProps(props);
    let { value: // These disabled props are handled by the state hook
    value , defaultValue: defaultValue , onChange: onChange , ...otherProps } = props;
    let state = $aAzec$reactstatelycolor.useColorFieldState(props);
    let inputRef = $aAzec$react.useRef();
    let { labelProps: labelProps , inputProps: inputProps  } = $aAzec$reactariacolor.useColorField(otherProps, state, inputRef);
    return(/*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).createElement($aAzec$reactspectrumtextfield.TextFieldBase, {
        ...otherProps,
        ref: ref,
        inputRef: inputRef,
        labelProps: labelProps,
        inputProps: inputProps,
        inputClassName: $aAzec$reactspectrumutils.classNames((/*@__PURE__*/$parcel$interopDefault($ddec4d6aa21146c4$exports)), 'react-spectrum-ColorField-input')
    }));
}
/**
 * ColorFields allow users to enter a color in #rrggbb hexadecimal format.
 */ const $8bdb9bc27ff3debb$export$b865d4358897bb17 = /*#__PURE__*/ ($parcel$interopDefault($aAzec$react)).forwardRef($8bdb9bc27ff3debb$var$ColorField);


$parcel$exportWildcard(module.exports, $5f016e950c2dc3a6$exports);
$parcel$exportWildcard(module.exports, $63a2864687444ae1$exports);
$parcel$exportWildcard(module.exports, $4537dec0de08c277$exports);
$parcel$exportWildcard(module.exports, $8bdb9bc27ff3debb$exports);
