// @react-spectrum/color/dist/module.js
import "./main.css";
import {dimensionValue as $kF0Zw$dimensionValue, useStyleProps as $kF0Zw$useStyleProps, useFocusableRef as $kF0Zw$useFocusableRef, classNames as $kF0Zw$classNames} from "@react-spectrum/utils";
import {mergeProps as $kF0Zw$mergeProps, useId as $kF0Zw$useId, useResizeObserver as $kF0Zw$useResizeObserver} from "@react-aria/utils";
import $kF0Zw$react, {useRef as $kF0Zw$useRef, useState as $kF0Zw$useState, useCallback as $kF0Zw$useCallback, useEffect as $kF0Zw$useEffect} from "react";
import {useColorArea as $kF0Zw$useColorArea, useColorWheel as $kF0Zw$useColorWheel, useColorSlider as $kF0Zw$useColorSlider, useColorField as $kF0Zw$useColorField} from "@react-aria/color";
import {useColorAreaState as $kF0Zw$useColorAreaState, useColorWheelState as $kF0Zw$useColorWheelState, useColorSliderState as $kF0Zw$useColorSliderState, useColorFieldState as $kF0Zw$useColorFieldState} from "@react-stately/color";
import {useFocusRing as $kF0Zw$useFocusRing} from "@react-aria/focus";
import {useProviderProps as $kF0Zw$useProviderProps} from "@react-spectrum/provider";
import {Flex as $kF0Zw$Flex} from "@react-spectrum/layout";
import {Label as $kF0Zw$Label} from "@react-spectrum/label";
import {useFocusVisible as $kF0Zw$useFocusVisible, useFocus as $kF0Zw$useFocus} from "@react-aria/interactions";
import {useLocale as $kF0Zw$useLocale} from "@react-aria/i18n";
import {TextFieldBase as $kF0Zw$TextFieldBase} from "@react-spectrum/textfield";

function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}
function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}
var $40046aa1a7ccb226$exports = {};

$parcel$export($40046aa1a7ccb226$exports, "ColorArea", () => $40046aa1a7ccb226$export$b2103f68a961418e);



var $9d103974e3e19727$exports = {};

$parcel$export($9d103974e3e19727$exports, "spectrum-ColorHandle", () => $9d103974e3e19727$export$cd9afaa621b6216f, (v) => $9d103974e3e19727$export$cd9afaa621b6216f = v);
$parcel$export($9d103974e3e19727$exports, "is-focused", () => $9d103974e3e19727$export$e7dc768d35940237, (v) => $9d103974e3e19727$export$e7dc768d35940237 = v);
$parcel$export($9d103974e3e19727$exports, "focus-ring", () => $9d103974e3e19727$export$f39a09f249340e2a, (v) => $9d103974e3e19727$export$f39a09f249340e2a = v);
$parcel$export($9d103974e3e19727$exports, "is-disabled", () => $9d103974e3e19727$export$d35bc1e505d1ebbf, (v) => $9d103974e3e19727$export$d35bc1e505d1ebbf = v);
$parcel$export($9d103974e3e19727$exports, "spectrum-ColorHandle-color", () => $9d103974e3e19727$export$afe4c366ed4e659c, (v) => $9d103974e3e19727$export$afe4c366ed4e659c = v);
var $9d103974e3e19727$export$cd9afaa621b6216f;
var $9d103974e3e19727$export$e7dc768d35940237;
var $9d103974e3e19727$export$f39a09f249340e2a;
var $9d103974e3e19727$export$d35bc1e505d1ebbf;
var $9d103974e3e19727$export$afe4c366ed4e659c;
$9d103974e3e19727$export$cd9afaa621b6216f = "spectrum-ColorHandle_5a9f41";
$9d103974e3e19727$export$e7dc768d35940237 = "is-focused_5a9f41";
$9d103974e3e19727$export$f39a09f249340e2a = "focus-ring_5a9f41";
$9d103974e3e19727$export$d35bc1e505d1ebbf = "is-disabled_5a9f41";
$9d103974e3e19727$export$afe4c366ed4e659c = "spectrum-ColorHandle-color_5a9f41";


var $f3cdcbdb4b87f010$exports = {};

$parcel$export($f3cdcbdb4b87f010$exports, "spectrum-ColorLoupe", () => $f3cdcbdb4b87f010$export$88d6ed2576b21809, (v) => $f3cdcbdb4b87f010$export$88d6ed2576b21809 = v);
$parcel$export($f3cdcbdb4b87f010$exports, "is-open", () => $f3cdcbdb4b87f010$export$a9781837241c946d, (v) => $f3cdcbdb4b87f010$export$a9781837241c946d = v);
$parcel$export($f3cdcbdb4b87f010$exports, "spectrum-ColorLoupe-outer", () => $f3cdcbdb4b87f010$export$d6ad38291e8aca9c, (v) => $f3cdcbdb4b87f010$export$d6ad38291e8aca9c = v);
$parcel$export($f3cdcbdb4b87f010$exports, "spectrum-ColorLoupe-inner-background", () => $f3cdcbdb4b87f010$export$397af82b48750fe5, (v) => $f3cdcbdb4b87f010$export$397af82b48750fe5 = v);
$parcel$export($f3cdcbdb4b87f010$exports, "spectrum-ColorLoupe-inner-checker", () => $f3cdcbdb4b87f010$export$f9a2a02e0b271024, (v) => $f3cdcbdb4b87f010$export$f9a2a02e0b271024 = v);
var $f3cdcbdb4b87f010$export$88d6ed2576b21809;
var $f3cdcbdb4b87f010$export$a9781837241c946d;
var $f3cdcbdb4b87f010$export$d6ad38291e8aca9c;
var $f3cdcbdb4b87f010$export$397af82b48750fe5;
var $f3cdcbdb4b87f010$export$f9a2a02e0b271024;
$f3cdcbdb4b87f010$export$88d6ed2576b21809 = "spectrum-ColorLoupe_c818a8";
$f3cdcbdb4b87f010$export$a9781837241c946d = "is-open_c818a8";
$f3cdcbdb4b87f010$export$d6ad38291e8aca9c = "spectrum-ColorLoupe-outer_c818a8";
$f3cdcbdb4b87f010$export$397af82b48750fe5 = "spectrum-ColorLoupe-inner-background_c818a8";
$f3cdcbdb4b87f010$export$f9a2a02e0b271024 = "spectrum-ColorLoupe-inner-checker_c818a8";



function $3a90e89d07d7bbd5$export$a3cc47cee1c1ccc(props) {
    let { value: value , isDisabled: isDisabled , isDragging: isDragging , isFocused: isFocused , children: children , className: className = '' , ...otherProps } = props;
    let patternId = $kF0Zw$useId();
    let valueCSS = value.toString('css');
    return(/*#__PURE__*/ $kF0Zw$react.createElement("div", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($9d103974e3e19727$exports)), 'spectrum-ColorHandle', {
            'is-focused': isFocused,
            'is-disabled': isDisabled
        }) + ' ' + className,
        ...otherProps
    }, /*#__PURE__*/ $kF0Zw$react.createElement("div", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($9d103974e3e19727$exports)), 'spectrum-ColorHandle-color'),
        style: {
            backgroundColor: valueCSS
        }
    }), /*#__PURE__*/ $kF0Zw$react.createElement("svg", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe', {
            'is-open': isDragging
        }),
        "aria-hidden": "true"
    }, /*#__PURE__*/ $kF0Zw$react.createElement("pattern", {
        id: patternId,
        x: "0",
        y: "0",
        width: "16",
        height: "16",
        patternUnits: "userSpaceOnUse"
    }, /*#__PURE__*/ $kF0Zw$react.createElement("rect", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe-inner-background'),
        x: "0",
        y: "0",
        width: "16",
        height: "16"
    }), /*#__PURE__*/ $kF0Zw$react.createElement("rect", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe-inner-checker'),
        x: "0",
        y: "0",
        width: "8",
        height: "8"
    }), /*#__PURE__*/ $kF0Zw$react.createElement("rect", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe-inner-checker'),
        x: "8",
        y: "8",
        width: "8",
        height: "8"
    })), /*#__PURE__*/ $kF0Zw$react.createElement("path", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe-inner'),
        d: "M25 1a24 24 0 0124 24c0 16.255-24 40-24 40S1 41.255 1 25A24 24 0 0125 1z",
        fill: `url(#${patternId})`
    }), /*#__PURE__*/ $kF0Zw$react.createElement("path", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe-inner'),
        d: "M25 1a24 24 0 0124 24c0 16.255-24 40-24 40S1 41.255 1 25A24 24 0 0125 1z",
        fill: valueCSS
    }), /*#__PURE__*/ $kF0Zw$react.createElement("path", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f3cdcbdb4b87f010$exports)), 'spectrum-ColorLoupe-outer'),
        d: "M25 3A21.98 21.98 0 003 25c0 6.2 4 14.794 11.568 24.853A144.233 144.233 0 0025 62.132a144.085 144.085 0 0010.4-12.239C42.99 39.816 47 31.209 47 25A21.98 21.98 0 0025 3m0-2a24 24 0 0124 24c0 16.255-24 40-24 40S1 41.255 1 25A24 24 0 0125 1z"
    })), children));
}




var $7d12200010f0192e$exports = {};

$parcel$export($7d12200010f0192e$exports, "spectrum-ColorArea", () => $7d12200010f0192e$export$bb6257a55a3c1efc, (v) => $7d12200010f0192e$export$bb6257a55a3c1efc = v);
$parcel$export($7d12200010f0192e$exports, "is-focused", () => $7d12200010f0192e$export$e7dc768d35940237, (v) => $7d12200010f0192e$export$e7dc768d35940237 = v);
$parcel$export($7d12200010f0192e$exports, "focus-ring", () => $7d12200010f0192e$export$f39a09f249340e2a, (v) => $7d12200010f0192e$export$f39a09f249340e2a = v);
$parcel$export($7d12200010f0192e$exports, "spectrum-ColorArea-handle", () => $7d12200010f0192e$export$d8addbd273c4e00, (v) => $7d12200010f0192e$export$d8addbd273c4e00 = v);
$parcel$export($7d12200010f0192e$exports, "is-disabled", () => $7d12200010f0192e$export$d35bc1e505d1ebbf, (v) => $7d12200010f0192e$export$d35bc1e505d1ebbf = v);
$parcel$export($7d12200010f0192e$exports, "spectrum-ColorArea-gradient", () => $7d12200010f0192e$export$40686f4fcb8a9916, (v) => $7d12200010f0192e$export$40686f4fcb8a9916 = v);
$parcel$export($7d12200010f0192e$exports, "spectrum-ColorArea-slider", () => $7d12200010f0192e$export$7d727cacaa7cea1e, (v) => $7d12200010f0192e$export$7d727cacaa7cea1e = v);
$parcel$export($7d12200010f0192e$exports, "spectrum-ColorHandle-color", () => $7d12200010f0192e$export$afe4c366ed4e659c, (v) => $7d12200010f0192e$export$afe4c366ed4e659c = v);
var $7d12200010f0192e$export$bb6257a55a3c1efc;
var $7d12200010f0192e$export$e7dc768d35940237;
var $7d12200010f0192e$export$f39a09f249340e2a;
var $7d12200010f0192e$export$d8addbd273c4e00;
var $7d12200010f0192e$export$d35bc1e505d1ebbf;
var $7d12200010f0192e$export$40686f4fcb8a9916;
var $7d12200010f0192e$export$7d727cacaa7cea1e;
var $7d12200010f0192e$export$afe4c366ed4e659c;
$7d12200010f0192e$export$bb6257a55a3c1efc = "spectrum-ColorArea_35e2c0";
$7d12200010f0192e$export$e7dc768d35940237 = "is-focused_35e2c0";
$7d12200010f0192e$export$f39a09f249340e2a = "focus-ring_35e2c0";
$7d12200010f0192e$export$d8addbd273c4e00 = "spectrum-ColorArea-handle_35e2c0";
$7d12200010f0192e$export$d35bc1e505d1ebbf = "is-disabled_35e2c0";
$7d12200010f0192e$export$40686f4fcb8a9916 = "spectrum-ColorArea-gradient_35e2c0";
$7d12200010f0192e$export$7d727cacaa7cea1e = "spectrum-ColorArea-slider_35e2c0";
$7d12200010f0192e$export$afe4c366ed4e659c = "spectrum-ColorHandle-color_35e2c0";






function $40046aa1a7ccb226$var$ColorArea(props, ref) {
    props = $kF0Zw$useProviderProps(props);
    let { isDisabled: isDisabled  } = props;
    let size = props.size && $kF0Zw$dimensionValue(props.size);
    let { styleProps: styleProps  } = $kF0Zw$useStyleProps(props);
    let inputXRef = $kF0Zw$useRef();
    let inputYRef = $kF0Zw$useRef();
    let containerRef = $kF0Zw$useFocusableRef(ref, inputXRef);
    let state = $kF0Zw$useColorAreaState(props);
    let { colorAreaProps: colorAreaProps , gradientProps: gradientProps , xInputProps: xInputProps , yInputProps: yInputProps , thumbProps: thumbProps  } = $kF0Zw$useColorArea({
        ...props,
        inputXRef: inputXRef,
        inputYRef: inputYRef,
        containerRef: containerRef
    }, state);
    let { focusProps: focusProps , isFocusVisible: isFocusVisible  } = $kF0Zw$useFocusRing();
    return(/*#__PURE__*/ $kF0Zw$react.createElement("div", {
        ...colorAreaProps,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($7d12200010f0192e$exports)), 'spectrum-ColorArea', {
            'is-disabled': isDisabled
        }, styleProps.className),
        ref: containerRef,
        style: {
            ...colorAreaProps.style,
            ...styleProps.style,
            // Workaround around https://github.com/adobe/spectrum-css/issues/1032
            width: size,
            height: size
        }
    }, /*#__PURE__*/ $kF0Zw$react.createElement("div", {
        ...gradientProps,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($7d12200010f0192e$exports)), 'spectrum-ColorArea-gradient')
    }), /*#__PURE__*/ $kF0Zw$react.createElement($3a90e89d07d7bbd5$export$a3cc47cee1c1ccc, {
        value: state.getDisplayColor(),
        isFocused: isFocusVisible,
        isDisabled: isDisabled,
        isDragging: state.isDragging,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($7d12200010f0192e$exports)), 'spectrum-ColorArea-handle'),
        ...thumbProps
    }, /*#__PURE__*/ $kF0Zw$react.createElement("div", {
        role: "presentation"
    }, /*#__PURE__*/ $kF0Zw$react.createElement("input", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($7d12200010f0192e$exports)), 'spectrum-ColorArea-slider'),
        ...$kF0Zw$mergeProps(xInputProps, focusProps),
        ref: inputXRef
    }), /*#__PURE__*/ $kF0Zw$react.createElement("input", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($7d12200010f0192e$exports)), 'spectrum-ColorArea-slider'),
        ...$kF0Zw$mergeProps(yInputProps, focusProps),
        ref: inputYRef
    })))));
}
let $40046aa1a7ccb226$export$b2103f68a961418e = /*#__PURE__*/ $kF0Zw$react.forwardRef($40046aa1a7ccb226$var$ColorArea);


var $3aac1c27203f04b5$exports = {};

$parcel$export($3aac1c27203f04b5$exports, "ColorWheel", () => $3aac1c27203f04b5$export$f80663f808113381);



var $f60625fffea2b12e$exports = {};

$parcel$export($f60625fffea2b12e$exports, "spectrum-ColorWheel", () => $f60625fffea2b12e$export$b70a5ca28702fbc6, (v) => $f60625fffea2b12e$export$b70a5ca28702fbc6 = v);
$parcel$export($f60625fffea2b12e$exports, "is-focused", () => $f60625fffea2b12e$export$e7dc768d35940237, (v) => $f60625fffea2b12e$export$e7dc768d35940237 = v);
$parcel$export($f60625fffea2b12e$exports, "spectrum-ColorWheel-handle", () => $f60625fffea2b12e$export$4cd23920eb8ac056, (v) => $f60625fffea2b12e$export$4cd23920eb8ac056 = v);
$parcel$export($f60625fffea2b12e$exports, "spectrum-ColorWheel-slider", () => $f60625fffea2b12e$export$f476ddda713d62b, (v) => $f60625fffea2b12e$export$f476ddda713d62b = v);
$parcel$export($f60625fffea2b12e$exports, "is-disabled", () => $f60625fffea2b12e$export$d35bc1e505d1ebbf, (v) => $f60625fffea2b12e$export$d35bc1e505d1ebbf = v);
$parcel$export($f60625fffea2b12e$exports, "is-dragged", () => $f60625fffea2b12e$export$8778c911bed6c759, (v) => $f60625fffea2b12e$export$8778c911bed6c759 = v);
$parcel$export($f60625fffea2b12e$exports, "spectrum-ColorWheel-gradient", () => $f60625fffea2b12e$export$9afaa791d7500a9b, (v) => $f60625fffea2b12e$export$9afaa791d7500a9b = v);
$parcel$export($f60625fffea2b12e$exports, "spectrum-ColorWheel-segment", () => $f60625fffea2b12e$export$ada9fbffdaa4f10, (v) => $f60625fffea2b12e$export$ada9fbffdaa4f10 = v);
var $f60625fffea2b12e$export$b70a5ca28702fbc6;
var $f60625fffea2b12e$export$e7dc768d35940237;
var $f60625fffea2b12e$export$4cd23920eb8ac056;
var $f60625fffea2b12e$export$f476ddda713d62b;
var $f60625fffea2b12e$export$d35bc1e505d1ebbf;
var $f60625fffea2b12e$export$8778c911bed6c759;
var $f60625fffea2b12e$export$9afaa791d7500a9b;
var $f60625fffea2b12e$export$ada9fbffdaa4f10;
$f60625fffea2b12e$export$b70a5ca28702fbc6 = "spectrum-ColorWheel_31462a";
$f60625fffea2b12e$export$e7dc768d35940237 = "is-focused_31462a";
$f60625fffea2b12e$export$4cd23920eb8ac056 = "spectrum-ColorWheel-handle_31462a";
$f60625fffea2b12e$export$f476ddda713d62b = "spectrum-ColorWheel-slider_31462a";
$f60625fffea2b12e$export$d35bc1e505d1ebbf = "is-disabled_31462a";
$f60625fffea2b12e$export$8778c911bed6c759 = "is-dragged_31462a";
$f60625fffea2b12e$export$9afaa791d7500a9b = "spectrum-ColorWheel-gradient_31462a";
$f60625fffea2b12e$export$ada9fbffdaa4f10 = "spectrum-ColorWheel-segment_31462a";







const $3aac1c27203f04b5$var$WHEEL_THICKNESS = 24;
function $3aac1c27203f04b5$var$ColorWheel(props, ref) {
    props = $kF0Zw$useProviderProps(props);
    let { isDisabled: isDisabled  } = props;
    let size = props.size && $kF0Zw$dimensionValue(props.size);
    let { styleProps: styleProps  } = $kF0Zw$useStyleProps(props);
    let inputRef = $kF0Zw$useRef(null);
    let containerRef = $kF0Zw$useFocusableRef(ref, inputRef);
    let [wheelRadius, setWheelRadius] = $kF0Zw$useState(null);
    let [wheelThickness, setWheelThickness] = $kF0Zw$useState($3aac1c27203f04b5$var$WHEEL_THICKNESS);
    let resizeHandler = $kF0Zw$useCallback(()=>{
        if (containerRef.current) {
            setWheelRadius(containerRef.current.offsetWidth / 2);
            let thickness = window.getComputedStyle(containerRef.current).getPropertyValue('--spectrum-colorwheel-track-thickness');
            if (thickness) setWheelThickness(parseInt(thickness, 10));
        }
    }, [
        containerRef,
        setWheelRadius,
        setWheelThickness
    ]);
    $kF0Zw$useEffect(()=>{
        // the size observer's fallback to the window resize event doesn't fire on mount
        if (wheelRadius == null) resizeHandler();
    }, [
        wheelRadius,
        resizeHandler
    ]);
    $kF0Zw$useResizeObserver({
        ref: containerRef,
        onResize: resizeHandler
    });
    let state = $kF0Zw$useColorWheelState(props);
    let { trackProps: trackProps , inputProps: inputProps , thumbProps: thumbProps  } = $kF0Zw$useColorWheel({
        ...props,
        innerRadius: wheelRadius - wheelThickness,
        outerRadius: wheelRadius
    }, state, inputRef);
    let { focusProps: focusProps , isFocusVisible: isFocusVisible  } = $kF0Zw$useFocusRing();
    return(/*#__PURE__*/ $kF0Zw$react.createElement("div", {
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f60625fffea2b12e$exports)), 'spectrum-ColorWheel', {
            'is-disabled': isDisabled
        }, styleProps.className),
        ref: containerRef,
        style: {
            ...styleProps.style,
            // Workaround around https://github.com/adobe/spectrum-css/issues/1032
            // @ts-ignore
            'width': size,
            'height': size
        }
    }, /*#__PURE__*/ $kF0Zw$react.createElement("div", {
        ...trackProps,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f60625fffea2b12e$exports)), 'spectrum-ColorWheel-gradient')
    }), /*#__PURE__*/ $kF0Zw$react.createElement($3a90e89d07d7bbd5$export$a3cc47cee1c1ccc, {
        value: state.getDisplayColor(),
        isFocused: isFocusVisible,
        isDisabled: isDisabled,
        isDragging: state.isDragging,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f60625fffea2b12e$exports)), 'spectrum-ColorWheel-handle'),
        ...thumbProps
    }, /*#__PURE__*/ $kF0Zw$react.createElement("input", {
        ...focusProps,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($f60625fffea2b12e$exports)), 'spectrum-ColorWheel-slider'),
        ...inputProps,
        ref: inputRef
    }))));
}
/**
 * ColorWheels allow users to adjust the hue of an HSL or HSB color value on a circular track.
 */ let $3aac1c27203f04b5$export$f80663f808113381 = /*#__PURE__*/ $kF0Zw$react.forwardRef($3aac1c27203f04b5$var$ColorWheel);


var $3870e4ce67ed7ee1$exports = {};

$parcel$export($3870e4ce67ed7ee1$exports, "ColorSlider", () => $3870e4ce67ed7ee1$export$44fd664bcca5b6fb);





var $670e80c7e5915fb5$exports = {};

$parcel$export($670e80c7e5915fb5$exports, "spectrum-ColorSlider", () => $670e80c7e5915fb5$export$a5f8b0989b27a604, (v) => $670e80c7e5915fb5$export$a5f8b0989b27a604 = v);
$parcel$export($670e80c7e5915fb5$exports, "is-focused", () => $670e80c7e5915fb5$export$e7dc768d35940237, (v) => $670e80c7e5915fb5$export$e7dc768d35940237 = v);
$parcel$export($670e80c7e5915fb5$exports, "spectrum-ColorSlider-handle", () => $670e80c7e5915fb5$export$8c273ba92fabe1f1, (v) => $670e80c7e5915fb5$export$8c273ba92fabe1f1 = v);
$parcel$export($670e80c7e5915fb5$exports, "spectrum-ColorSlider-slider", () => $670e80c7e5915fb5$export$2f08b72944af536, (v) => $670e80c7e5915fb5$export$2f08b72944af536 = v);
$parcel$export($670e80c7e5915fb5$exports, "spectrum-ColorSlider-container--horizontal", () => $670e80c7e5915fb5$export$33e97914f6746614, (v) => $670e80c7e5915fb5$export$33e97914f6746614 = v);
$parcel$export($670e80c7e5915fb5$exports, "spectrum-ColorSlider-container--vertical", () => $670e80c7e5915fb5$export$cd1c093289104c66, (v) => $670e80c7e5915fb5$export$cd1c093289104c66 = v);
$parcel$export($670e80c7e5915fb5$exports, "is-disabled", () => $670e80c7e5915fb5$export$d35bc1e505d1ebbf, (v) => $670e80c7e5915fb5$export$d35bc1e505d1ebbf = v);
$parcel$export($670e80c7e5915fb5$exports, "spectrum-ColorSlider--vertical", () => $670e80c7e5915fb5$export$83dcf61b398bfaf1, (v) => $670e80c7e5915fb5$export$83dcf61b398bfaf1 = v);
var $670e80c7e5915fb5$export$a5f8b0989b27a604;
var $670e80c7e5915fb5$export$e7dc768d35940237;
var $670e80c7e5915fb5$export$8c273ba92fabe1f1;
var $670e80c7e5915fb5$export$2f08b72944af536;
var $670e80c7e5915fb5$export$33e97914f6746614;
var $670e80c7e5915fb5$export$cd1c093289104c66;
var $670e80c7e5915fb5$export$d35bc1e505d1ebbf;
var $670e80c7e5915fb5$export$83dcf61b398bfaf1;
$670e80c7e5915fb5$export$a5f8b0989b27a604 = "spectrum-ColorSlider_a5bb06";
$670e80c7e5915fb5$export$e7dc768d35940237 = "is-focused_a5bb06";
$670e80c7e5915fb5$export$8c273ba92fabe1f1 = "spectrum-ColorSlider-handle_a5bb06";
$670e80c7e5915fb5$export$2f08b72944af536 = "spectrum-ColorSlider-slider_a5bb06";
$670e80c7e5915fb5$export$33e97914f6746614 = "spectrum-ColorSlider-container--horizontal_a5bb06";
$670e80c7e5915fb5$export$cd1c093289104c66 = "spectrum-ColorSlider-container--vertical_a5bb06";
$670e80c7e5915fb5$export$d35bc1e505d1ebbf = "is-disabled_a5bb06";
$670e80c7e5915fb5$export$83dcf61b398bfaf1 = "spectrum-ColorSlider--vertical_a5bb06";







function $3870e4ce67ed7ee1$var$ColorSlider(props, ref) {
    props = $kF0Zw$useProviderProps(props);
    let { isDisabled: isDisabled , channel: channel , orientation: orientation , label: label , showValueLabel: showValueLabel , 'aria-label': ariaLabel  } = props;
    let vertical = orientation === 'vertical';
    let { styleProps: styleProps  } = $kF0Zw$useStyleProps(props);
    let { locale: locale  } = $kF0Zw$useLocale();
    let inputRef = $kF0Zw$useRef();
    let trackRef = $kF0Zw$useRef();
    let domRef = $kF0Zw$useFocusableRef(ref, inputRef);
    let state = $kF0Zw$useColorSliderState({
        ...props,
        locale: locale
    });
    // If vertical and a label is provided, use it as an aria-label instead.
    if (vertical && label) {
        ariaLabel = ariaLabel || (typeof label === 'string' ? label : null);
        label = null;
    }
    // If no external label, aria-label or aria-labelledby is provided,
    // default to displaying the localized channel value.
    // Specifically check if label is undefined. If label is `null` then display no visible label.
    // A default aria-label is provided by useColorSlider in that case.
    if (label === undefined && !ariaLabel && !props['aria-labelledby'] && !vertical) label = state.value.getChannelName(channel, locale);
    // Show the value label by default if there is a visible label
    if (showValueLabel == null) showValueLabel = !!label;
    let { inputProps: inputProps , thumbProps: thumbProps , trackProps: trackProps , labelProps: labelProps , outputProps: outputProps  } = $kF0Zw$useColorSlider({
        ...props,
        label: label,
        'aria-label': ariaLabel,
        trackRef: trackRef,
        inputRef: inputRef
    }, state);
    let { isFocusVisible: isFocusVisible  } = $kF0Zw$useFocusVisible();
    let [isFocused, setIsFocused] = $kF0Zw$useState(false);
    let { focusProps: focusProps  } = $kF0Zw$useFocus({
        isDisabled: isDisabled,
        onFocusChange: setIsFocused
    });
    let alignLabel;
    if (vertical) alignLabel = 'center';
    else if (label != null && showValueLabel) alignLabel = 'space-between';
    else if (label != null) alignLabel = 'flex-start';
    else if (showValueLabel) alignLabel = 'flex-end';
    return(/*#__PURE__*/ $kF0Zw$react.createElement("div", {
        ref: domRef,
        ...styleProps,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($670e80c7e5915fb5$exports)), {
            'spectrum-ColorSlider-container--horizontal': !vertical,
            'spectrum-ColorSlider-container--vertical': vertical
        })
    }, label && /*#__PURE__*/ $kF0Zw$react.createElement($kF0Zw$Flex, {
        direction: "row",
        justifyContent: alignLabel
    }, /*#__PURE__*/ $kF0Zw$react.createElement($kF0Zw$Label, labelProps, label), showValueLabel && /*#__PURE__*/ $kF0Zw$react.createElement($kF0Zw$Label, {
        elementType: "span"
    }, /*#__PURE__*/ $kF0Zw$react.createElement("output", outputProps, state.value.formatChannelValue(channel, locale)))), /*#__PURE__*/ $kF0Zw$react.createElement("div", {
        ...trackProps,
        ref: trackRef,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($670e80c7e5915fb5$exports)), 'spectrum-ColorSlider', {
            'is-disabled': isDisabled,
            'spectrum-ColorSlider--vertical': vertical
        })
    }, /*#__PURE__*/ $kF0Zw$react.createElement($3a90e89d07d7bbd5$export$a3cc47cee1c1ccc, {
        value: state.getDisplayColor(),
        isFocused: isFocused && isFocusVisible,
        isDisabled: isDisabled,
        isDragging: state.isThumbDragging(0),
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($670e80c7e5915fb5$exports)), 'spectrum-ColorSlider-handle'),
        ...thumbProps
    }, /*#__PURE__*/ $kF0Zw$react.createElement("input", {
        ...inputProps,
        ...focusProps,
        ref: inputRef,
        className: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($670e80c7e5915fb5$exports)), 'spectrum-ColorSlider-slider')
    })))));
}
/**
 * ColorSliders allow users to adjust an individual channel of a color value.
 */ let $3870e4ce67ed7ee1$export$44fd664bcca5b6fb = /*#__PURE__*/ $kF0Zw$react.forwardRef($3870e4ce67ed7ee1$var$ColorSlider);


var $0077d47e004328f0$exports = {};

$parcel$export($0077d47e004328f0$exports, "ColorField", () => $0077d47e004328f0$export$b865d4358897bb17);


var $9d47b39540b9e48a$exports = {};

$parcel$export($9d47b39540b9e48a$exports, "react-spectrum-ColorField-input", () => $9d47b39540b9e48a$export$7d80c6630a750b1, (v) => $9d47b39540b9e48a$export$7d80c6630a750b1 = v);
var $9d47b39540b9e48a$export$7d80c6630a750b1;
$9d47b39540b9e48a$export$7d80c6630a750b1 = "react-spectrum-ColorField-input_7bde3a";






function $0077d47e004328f0$var$ColorField(props, ref) {
    props = $kF0Zw$useProviderProps(props);
    let { value: // These disabled props are handled by the state hook
    value , defaultValue: defaultValue , onChange: onChange , ...otherProps } = props;
    let state = $kF0Zw$useColorFieldState(props);
    let inputRef = $kF0Zw$useRef();
    let { labelProps: labelProps , inputProps: inputProps  } = $kF0Zw$useColorField(otherProps, state, inputRef);
    return(/*#__PURE__*/ $kF0Zw$react.createElement($kF0Zw$TextFieldBase, {
        ...otherProps,
        ref: ref,
        inputRef: inputRef,
        labelProps: labelProps,
        inputProps: inputProps,
        inputClassName: $kF0Zw$classNames((/*@__PURE__*/$parcel$interopDefault($9d47b39540b9e48a$exports)), 'react-spectrum-ColorField-input')
    }));
}
/**
 * ColorFields allow users to enter a color in #rrggbb hexadecimal format.
 */ const $0077d47e004328f0$export$b865d4358897bb17 = /*#__PURE__*/ $kF0Zw$react.forwardRef($0077d47e004328f0$var$ColorField);




export {$40046aa1a7ccb226$export$b2103f68a961418e as ColorArea, $3aac1c27203f04b5$export$f80663f808113381 as ColorWheel, $3870e4ce67ed7ee1$export$44fd664bcca5b6fb as ColorSlider, $0077d47e004328f0$export$b865d4358897bb17 as ColorField};
